nomos-mix/network/src/behaviour.rs

```diff
@@ -0,0 +1,231 @@
+        }
+    }
+
+    /// Handles an event generated by the [`MixConnectionHandler`]
+    /// dedicated to the connection identified by `peer_id` and `connection_id`.
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        event: THandlerOutEvent<Self>,
+    ) {
+        match event {
+            // A message was forwarded from the peer.
+            ToBehaviour::Message(message) => {
+                if self
+                    .duplicate_cache
+                    .cache_set(message_id(&message), ())
+                    .is_some()
+                {
+                    return;
+                }
+
+                // Try to unwrap the message.
+                match unwrap_message(&message) {
+                    Ok((unwrapped_msg, fully_unwrapped)) => {
+                        if fully_unwrapped {
+                            self.events.push_back(ToSwarm::GenerateEvent(
+                                Event::FullyUnwrappedMessage(unwrapped_msg),
+                            ));
+                        } else if let Err(e) = self.forward_message(unwrapped_msg, None) {
+                            tracing::error!("Failed to forward message: {:?}", e);
+                        }
+                    }
+                    Err(nomos_mix_message::Error::MsgUnwrapNotAllowed) => {
+                        // Forward the received message as it is.
+                        if let Err(e) = self.forward_message(message, Some(peer_id)) {
+                            tracing::error!("Failed to forward message: {:?}", e);
+                        }
+                    }
+                    Err(e) => {
+                        tracing::error!("Failed to unwrap message: {:?}", e);
+                    }
+                }
+            }
+            // The connection was fully negotiated by the peer,
+            // which means that the peer supports the mix protocol.
+            ToBehaviour::FullyNegotiatedOutbound => {
+                self.add_negotiated_peer(peer_id, connection_id);
+            }
+            ToBehaviour::NegotiationFailed => {
+                self.remove_negotiated_peer(&peer_id, &connection_id);
+            }
+            ToBehaviour::IOError(error) => {
+                // TODO: Consider removing the peer from the connected_peers and closing the connection
+                self.events
+                    .push_back(ToSwarm::GenerateEvent(Event::Error(Error::PeerIOError {
+                        error,
+                        peer_id,
+                        connection_id,
+                    })));
+            }
+        }
+    }
+
+    /// Polls for things that swarm should do.
+    fn poll(&mut self, _: &mut Context<'_>) -> Poll<ToSwarm<Self::ToSwarm, THandlerInEvent<Self>>> {
+        if let Some(event) = self.events.pop_front() {
+            Poll::Ready(event)
+        } else {
+            Poll::Pending
```

Comment:
Where do we schedule the waker for waking up?
In my previous experience this may be problematic. If you return pending without telling the waker when to reschedule for polling it may never be called again. Probably we need to store the waker. And when we push something to events we would want to call for the waker to schedule the task again.

Reply:
Oh, let me check it again. I saw that the waker is used in the DA implementation, but also saw that the `ping` and `gossipsub` in libp2p don't call the waker. Let me recheck it, and figure out the proper way.

Reply:
I just found that people have been complaining about the way how libp2p guys implement `NetworkBehaviour` and `ConnectionHandler` (e.g. [here](https://discuss.libp2p.io/t/implementing-custom-protocols-with-rust-libp2p/649)). I also found an [answer](https://github.com/libp2p/rust-libp2p/issues/5147#issuecomment-1937958768) from the rust-libp2p maintainer. He says that we don't need to call `Waker.wake()` if any change on the `self.events` is made by the API defined in the `NetworkBehaviour` trait because the `Swarm` polls the behaviour again if those APIs are called. That's why any protocol implementations in rust-libp2p (including `ping` and `gossipsub`) don't call `Waker.wake()`. 
In this PR, the data structures (like `self.events`) are updated only by trait APIs. However, I am not sure if the way that the `Swarm` uses is good. I think it can be problematic if the implementer updates the data structure via additional APIs and forgets to call `Waker.wake()`. 
What do you think about this issue? I'm thinking about calling `Waker.wake()` as you did for DA, which is the standard polling convention. But, also thinking about the following sentence that the rust-libp2p maintainer wrote:
> This has been done to avoid excessive use of Wakers.

Reply:
[a3b2a5b](https://github.com/logos-co/nomos-node/pull/765/commits/a3b2a5bc97f144188563c4ca31b5ccafcd1f7e96) I think it's much better to follow the standard convention, as you did, regardless of what rust-libp2p does behind the scene. Actually, the `wake` calls that I added is not so different from what the rust-libp2p guy said, because I call the `wake` at the end of `on_connection_handler_event` each time.

Reply:
> [a3b2a5b](https://github.com/logos-co/nomos-node/pull/765/commits/a3b2a5bc97f144188563c4ca31b5ccafcd1f7e96) I think it's much better to follow the standard convention, as you did, regardless of what rust-libp2p does behind the scene. Actually, the `wake` calls that I added is not so different from what the rust-libp2p guy said, because I call the `wake` at the end of `on_connection_handler_event` each time.

I think is better. Mostly because what they claim to poll is not true. I encountered issues when not calling `wake` in where the future is never polled again and it just stalls. In this case is really simple because is just storing the waker and calling wake whenever pushing an event.

Reply:
I agree. Good point. Thank you. Let’s go with the waker that I added